Add msbuild 16.6

mnme opened the pull request:
The current msbuild 15.4 recipe is broken. This pull request adds a new recipe for version 16.6. I was only able to test the msbuild-native recipe with mono 6.8.0.105 and poky zeus, every other configuration is completely untested.

I added a recipe for dotnet as it is a dependency of msbuild. I copied it from [intel-iot-devkit/meta-iot-cloud](https://github.com/intel-iot-devkit/meta-iot-cloud) as I didn't want to depend on the whole layer just for a single recipe. The layer is MIT licensed, this probably needs attribution.

There is also a change to use correct paths in the config for mono-native, I think this should not break other things.

It would be nice if someone can test this a bit more and give feedback.

Fixes #38.

ajlennon commented:
Thanks @mnme - it's great news we're going to have a working msbuild again!

zboszor commented:
Today, msbuild 15.4 didn't build for me because of some dependency download errors. I was guessing that MS may have archived some old dependencies that can't be downloaded anymore. So I was forced to test this PR finally.

I tried to apply the PR patch on meta-sicom over the master branch commit just before removing Sumo (Yocto 2.5) support because I still have to use it. The patch applied almost cleanly with a single fuzz.

The build failed first because there's no `unzip` command available. I added `unzip-native` to `DEPENDS` to fix it.

Then the build failed with a series of warnings and errors like this below for different URLs:
```
git/Directory.Build.props(3,3): warning : Unable to load the service index for source https://dotnet.myget.org/F/aspnetcore-release/api/v3/index.json.
git/artifacts/mono-msbuild/NuGet.targets(123,5): error : Unable to load the service index for source https://dotnet.myget.org/F/aspnetcore-release/api/v3/index.json.
git/artifacts/mono-msbuild/NuGet.targets(123,5): error :   404 (Feed does not exist.)
```
I also tried a newer `SRCREV = "94d0c55bc96f297618d50cc32167ddba9fee30b0"` from the linux-packaging-msbuild repository. It made the last patch unnecessary from the new msbuild 16.6 recipe. But the build still failed with the same series of 404 errors.

Maybe there's a generic download issue at MS at the moment that makes both msbuild recipe versions fail.

zboszor commented:
This seems to be a similar issue that I've hit: https://github.com/dotnet/source-build/issues/1971
The dotnet recipe should be 3.1.11 with SDK version 3.1.111, both were released 5 days ago.
The below lines need to be changed in the dotnet recipe: 
```
HOST_FXR = "3.1.11"
SHARED_FRAMEWORK = "3.1.11"
SDK = "3.1.111"

SRC_URI = "https://download.visualstudio.microsoft.com/download/pr/d5d940c0-4c2f-4cbb-8d12-33bfb30c4db8/619b5be6e995cead6e9432134f903711/${BPN}-sdk-${SDK}-linux-x64.tar.gz"

SRC_URI[sha256sum] = "a755b37aa328d160a39b39c0cd120ca7e318814ef024682bac49e75183952a89"
```
Despite these changes, msbuild-native still fails with the "404 (Feed does not exist.)" errors as it still references the old myget.org URLs.

mnme commented:
Thank you very much for testing the recipe! I added some local changes I had lying around. Additionally, I added a patch to fix the 404's you were seeing. We can revert that one as soon as the issue is resolved upstream.

zboszor commented on classes/mono.bbclass:
```diff
@@ -40,4 +40,5 @@ export NUGET_HTTP_CACHE_PATH="${WORKDIR}/mono-nuget-http-cache"
 
 do_configure_prepend() {
 	mkdir -p ${NUGET_PACKAGES} ${NUGET_HTTP_CACHE_PATH}
+    cert-sync --user ${STAGING_DIR_NATIVE}/etc/ssl/certs/ca-certificates.crt
```
Please use a TAB here, not spaces.

mnme replied:
Good catch, that one slipped trough

zboszor commented:
You can squash the dotnet patches together so the 3.1.11 version is introduced immediately without going with 3.1.8 first.
`git rebase -i` is your friend.

The cert-sync patch against `mono.bbclass` can be moved earlier in the series and the msbuild recipe patch doesn't need to add and then remove it.

The msbuild recipe still needs unzip-native in DEPENDS. I am not sure when Yocto added `unzip` to `HOSTTOOLS` or whether it's your local change, but building this recipe on Yocto 2.5 definitely needs it. It's better to have it just in case. Without this, I get:
```
Log data follows:
| DEBUG: Executing shell function do_compile
| ** Downloading MSBUILD from https://github.com/mono/msbuild/releases/download/0.08/mono_msbuild_6.4.0.208.zip
| ./eng/cibuild_bootstrapped_msbuild.sh: line 61: unzip: command not found
| mv: cannot stat ‘/data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/git/eng/../artifacts/msbuild’: No such file or directory
| cp: cannot create regular file ‘/data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/git/eng/../artifacts/mono-msbuild/SdkResolvers/Microsoft.DotNet.MSBuildSdkResolver/’: No such file or directory
| sed: can't read /data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/git/eng/../artifacts/mono-msbuild/msbuild: No such file or directory
| chmod: cannot access ‘/data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/git/eng/../artifacts/mono-msbuild/MSBuild.dll’: No such file or directory
| Cannot open assembly '/data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/git/eng/../artifacts/mono-msbuild/MSBuild.dll': No such file or directory.
| WARNING: /data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/temp/run.do_compile.29418:1 exit 2 from './eng/cibuild_bootstrapped_msbuild.sh --host_type mono --configuration Release --skip_tests /p:DisableNerdbankVersioning=true'
| ERROR: Function failed: do_compile (log file is located at /data/OE/sicom-manifest-2.5/build/tmp-sicom-glibc/work/x86_64-linux/msbuild-native/16.6-r0/temp/log.do_compile.29418)
```
But otherwise it looks good.

zboszor commented:
Another thing: this version of msbuild needs dotnet assemblies for version 4.7.2. It can be build only with Mono 5.18+.

mnme commented:
Okay, I cleaned the commits up. Some open questions before I push:

- Is there a preferred indentation style? Most of my changes use spaces currently, I could easily convert them to tabs if that fits the repo better.
- Regarding your last remark, is there a way to express dependencies on specific versions?

zboszor commented:
In the shell script functions of the recipe, TABs are preferred. With python based functions, indentation with 4 spaces is about mandatory. Yes, this potentially makes a recipe to use a mixed indentation style.

zboszor commented:
As for expressing a versioned dependency, I am not sure if the build-time dependency variable DEPENDS handle it.
RDEPENDS for runtime dependencies handle it this way:
```
RDEPENDS_${PN} = "foo (>= version)"
```

zboszor commented:
Anyway, I just tried to build msbuild-native on Yocto 2.5 using Mono 5.20.1.34 and it still failed with Roslyn errors.
Which version did you build it with? Any particular 6.x? 6.8.0.105 was added before Yocto 2.5 compatibility was removed.

mnme commented:
> In the shell script functions of the recipe, TABs are preferred. With python based functions, indentation with 4 spaces is about mandatory. Yes, this potentially makes a recipe to use a mixed indentation style.

And outside of both, e.g. when indenting `SRC_URI`s? Probably tabs, too?

> As for expressing a versioned dependency, I am not sure if the build-time dependency variable DEPENDS handle it.
> RDEPENDS for runtime dependencies handle it this way:
> 
> ```
> RDEPENDS_${PN} = "foo (>= version)"
> ```

Interesting, I didn't knew about that. The manual only mentions it in the `RDEPENDS` section, but I think it could work for `DEPENDS` too, as it is a bitbake feature. The problem that I currently face is that the mono dependency is declared in `mono.bbclass`, so I would need to override it.

> Anyway, I just tried to build msbuild-native on Yocto 2.5 using Mono 5.20.1.34 and it still failed with errors.
> Which version did you build it with? Any particular 6.x? 6.8.0.105 was added before Yocto 2.5 compatibility was removed.

I've only completely tested (i.e. building a C# application with the resulting msbuild) the recipe on zeus with mono-6.8.0.105. I'm currently using mono 6.10.0.104 to test if the patches build.

zboszor commented:
msbuild-native itself succeeded. The other software using msbuild for its build process is still in progress.
I will also re-test building msbuild with Mono 5.20 with this change adapted from the `mono-native-6.xx-base.inc` change in your PR:
```
diff --git a/recipes-mono/mono/mono-native-5.xx-base.inc b/recipes-mono/mono/mono-native-5.xx-base.inc
index 50460f1..02b3d52 100644
--- a/recipes-mono/mono/mono-native-5.xx-base.inc
+++ b/recipes-mono/mono/mono-native-5.xx-base.inc
@@ -16,3 +16,7 @@ mono_copy_libdir_mono() {
 }
 
 do_populate_sysroot[postfuncs] += " mono_copy_libdir_mono "
+
+do_install_append() {
+    sed "s|\$mono_libdir|${libdir}|g" -i ${D}${sysconfdir}/mono/config
+}
```

ajlennon commented:
Great stuff guys. Let me know when you're happy for final review for merge. Thanks @zboszor have give you triage access

mnme commented:
I've pushed what I have done so far. I'll probably clean it up a little more regarding coding standards, but otherwise it's finished.

ajlennon commented:
Thanks @mnme @zboszor ! Merged in...

zboszor commented:
I  am sorry I didn't report back before merging, I got distracted. With msbuild 16.6 and Mono 6.8.0.105, our in-house project builds successfully.

Unfortunately, the previously quoted patch against `mono-native-5.xx-base.inc` does not help, msbuild-native doesn't build successfully against mono-native 5.20.1.34.
